### PR TITLE
OAuth: Configurable user name attribute

### LINKED ISSUE #28224: OAuth: configurable user Name (displayName) attribute
**What would you like to be added**:
It's possible to configure email attribute name for OAuth users to use attribute other than default `Email` for user's email. It would be nice also to make user name attribute configurable as well (now grafana uses default `name` or `display_name` field).

**Why is this needed**:
This can make OAuth more flexible.

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What this PR does / why we need it**:
This PR adds configuration options for specifying custom attribute name or path for user name.

**Which issue(s) this PR fixes**:
This PR fixes #28224 

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/login/social/generic_oauth.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ type SocialGenericOAuth struct {
 	emailAttributeName   string
 	emailAttributePath   string
 	loginAttributePath   string
+	nameAttributeName    string
```
Comment (sakjur) [inline pkg/login/social/generic_oauth.go:27]: Why do we add both `nameAttributeName` and `nameAttributePath`? Isn't the `name` variant of other properties just a legacy setting from before we supported JMESPaths?

File: pkg/login/social/generic_oauth.go [hunk 2]
```diff
@@ -107,12 +109,9 @@ func (s *SocialGenericOAuth) UserInfo(client *http.Client, token *oauth2.Token)
...
 		if userInfo.Name == "" {
-			if data.Name != "" {
-				s.log.Debug("Setting user info name from name field")
-				userInfo.Name = data.Name
-			} else if data.DisplayName != "" {
-				s.log.Debug("Setting user info name from display name field")
-				userInfo.Name = data.DisplayName
+			userInfo.Name = s.extractUserName(data)
+			if userInfo.Name != "" {
+				s.log.Debug("Set user info name from extracted name", "name", userInfo.Name)
```
Comment (papagian) [inline pkg/login/social/generic_oauth.go:114]: I think that this debug message is redundant given that each special case in `extractUserName` is logged.

File: pkg/login/social/generic_oauth.go (hunk 1 above)
Comment (alexanderzobnin) [reply pkg/login/social/generic_oauth.go:27]: Hm, good question, I'll check it out. Looks like for login we only have `loginAttributePath` option.

**Review by papagian — APPROVED** [review]
LGTM, it would be nice to add some tests!

File: pkg/login/social/generic_oauth.go (hunk 1 above)
Comment (alexanderzobnin) [reply pkg/login/social/generic_oauth.go:27]: @marefr What do you think?

**Review by xlson — APPROVED** [review]

File: docs/sources/auth/generic-oauth.md [hunk 3]
```diff
@@ -80,6 +80,8 @@ Customize user login using `login_attribute_path` configuration option. Order of
 
 You can customize the attribute name used to extract the ID token from the returned OAuth token with the `id_token_attribute_name` option.
 
+Customize user's displayed name using `name_attribute_path` configuration option. Order of operations is the same as for the `login_attribute_path` option.
```
Comment (sakjur) [inline docs/sources/auth/generic-oauth.md:83]: Unless we'll squeeze this into 7.3, we should split out the docs changes to a separate PR that's merged together with 7.4-beta1 because our docs release pipeline publishes all docs changes immediately.

```suggestion
You can set the user's display name with JMESPath using the `name_attribute_path` configuration option. It operates the same way as the `login_attribute_path` option.

> **Note:** `name_attribute_path` is added in 7.4.
```

**Review by sakjur — APPROVED** [review]
I left a comment about the docs, we can either split docs out into a separate PR in which case this is :+1: from me, otherwise we should wait for a technical writer to give a LGTM from their perspective.

File: docs/sources/auth/generic-oauth.md (hunk 3 above)
Comment (alexanderzobnin) [reply docs/sources/auth/generic-oauth.md:83]: Yes, forget about this, good catch!

**Review by sakjur — APPROVED** [review]
